🚀 3단계 - 경로 조회 기능

hhhhhwi opened the pull request:
안녕하세요 용주님..! 
3단계 구현이 늦었습니다 ㅠ

인수테스트를 작성하고, 도메인을 설계 및 구현하는 방향인 Inside TDD 방식으로 사이클을 진행하다,
단위 테스트를 작성하며 api 구현을 진행하는 Outside TDD 방식으로 개발하려고 했으나..
해당 흐름으로는 제대로 진행을 못 했던 것 같습니다 ㅠ

테스트를 먼저 작성을 하다보니, 구현을 억지로..? 맞춘다는 느낌도 없지 않았습니다.

좀 늦게 PR을 올려 조급함에 부족한 점이 많을 것 같은데요 😢 
이번에도 피드백 잘 부탁드립니다..! 감사합니다!

hhhhhwi committed:
refactor: 가독성을 위해 변수명 변경

hhhhhwi committed:
docs: readme 추가 작성
-  경로 조회 시나리오 작성

hhhhhwi committed:
test: 경로 조회 인수 테스트 구현

hhhhhwi committed:
test: 경로 조회를 하는 도메인 path에 대한 단위 테스트 구현

hhhhhwi committed:
feat: 경로 조회를 하는 Path 구현
- jgrapht 라이브러리를 활용

hhhhhwi committed:
test: 경로 서비스 단위 테스트 작성

hhhhhwi committed:
feat: 경로 조회 api 구현 및 구간 추가 Path 객체에 경로를 추가하도록 수정

hhhhhwi committed:
feat: 없는 경로 조회 시 Exception 발생

hhhhhwi committed:
refactor: 가독성을 위해 최단 경로 조회 성공 메소드 생성

hhhhhwi committed:
refactor: 노선 및 구간 생성 시마다 경로가 저장되기 때문에 BeforeAll를 통해 Path 인수 테스트를 위한 객체 생성

hhhhhwi committed:
refactor: 존재하지 않는 역의 경로를 조회할 경우 exception 처리

hhhhhwi committed:
refactor: Path 인수 테스트를 위한 Fixture 생성

hhhhhwi committed:
refactor: Path에 대한 Unit 테스트와 내용이 겹쳐 삭제

hhhhhwi committed:
docs: 테스트 fixture 변경으로 인해 시나리오의 역 명 변경

hhhhhwi commented on src/test/java/nextstep/subway/utils/subway/PathFixture.java:
Sections.addSection 시 Path를 추가하다보니.. StationFixture 로는 (static 변수로 선언한 게 잘못인거 같기도 합니다 😭 ) 제대로 된 검증이 되지 않아 Path 테스트용의 FIxture를 생성했습니다..!
쓸데없이 Fixture가 계속 늘어나는 것 같아 좀 고민이고, 좀 더 효과적으로 Fixture 관리를 할 수 있는 방법이 없을까 고민이 많이 되었습니다..!

testrace replied:
다양한 관점이 있겠지만 fixture를 분리하면서 테스트간의 격리가 보장된다면 fixture가 많은게 크게 문제가 되진 않을 것 같아요 😃 
학습 과정이니 이런저런 시도를 해보시는게 좋은 것 같아요!

hhhhhwi commented on src/test/java/nextstep/subway/unit/PathServiceTest.java:
현재 PathService는 Path에서 조회한 값을 Dto로 변환하는 역할밖에 없는데요..! getPath를 테스트하자니 Path의 Unit 테스트와 다를 게 없다고 느껴져서 삭제했습니다..!
Service에 대한 Unit 테스트를 추가하는 게 좋을까요?!

hhhhhwi commented on src/main/java/nextstep/subway/domain/Path.java:
```diff
@@ -20,7 +20,13 @@ public static void addPath(Long source, Long target, int weight) {
+
+		try {
+			graphPath = dijkstraShortestPath.getPath(source, target);
+		} catch (IllegalArgumentException e) {
```
dijkstraShortestPath.getPath에서 던지는 Exception 으로 처리하고 싶어 try ~ catch를 사용했는데 코드가 좀 지저분해진 것 같네요 😞

testrace commented on README.md:
`````diff
@@ -111,8 +111,98 @@
+  ````
+
+## TODO
+- [x] 시나리오 작성
+  - 성공 시나리오
+    - Scenario: 경로가 다건일 경우, 최단 경로 조회 성공
+      ````
+      Given 상행역 : 도곡역 / 하행역 : 학여울역 / 길이 : 6                                 
+            상행역 : 학여울역 / 하행역 : 수서역 / 길이 : 10 인 구간들이 존재하는 노선을 생성한다.          
+      Given 상행역 : 도곡역 / 하행역 : 개포동역 / 길이 : 4                                 
+            상행역 : 개포동역 / 하행역 : 수서역 / 길이 : 8 인 구간들이 존재하는 노선을 생성한다.           
+      When 수서역부터 도곡역까지의 경로를 조회하면                                            
+      Then 경로에 있는 역 목록은 수서역, 개포동, 도곡역 순서대로 구성된다.                            
+      Then 구간의 거리는 12이다.                                                    
+      ````
`````
인수 조건 도출 👍

testrace commented on src/main/java/nextstep/subway/domain/Path.java:
```diff
@@ -0,0 +1,45 @@
+
+public final class Path {
+	private static WeightedMultigraph<Long, DefaultWeightedEdge> path = new WeightedMultigraph(DefaultWeightedEdge.class);
+	private static DijkstraShortestPath dijkstraShortestPath = new DijkstraShortestPath<>(path);
```
경로 조회를 위한 로직을 클래스로 분리해주셨네요 👍 
클래스에는 final, 다익스트라 변수에는 static이 없어도 기능에는 문제가 없을 것 같은데 붙여주신 이유가 있으실까요?

hhhhhwi replied:
구간 등록 및 수정을 하는 시점에 경로까지 수정해야 한다고 생각하여 final 클래스로 선언했던 건데요..! final 클래스로 선언하다보니 static 변수를 사용하는 등.. 어쩔 수 없는 코드를 작성하게 되버렸습니다.. 
용주님께서 제안주신 방향처럼 경로 조회 시점에 Path 관련 객체를 생성하도록 수정해보았습니다..!

testrace commented on src/main/java/nextstep/subway/domain/Path.java:
```diff
@@ -0,0 +1,45 @@
+
+	public static void removePath(Long source, Long target) {
+		path.removeEdge(source, target);
+	}
```
경로조회 이후 반드시 removePath() 메서드를 호출해줘야한다는 사실은 이 코드를 작성한 개발자 외엔 파악하기 쉽지 않을 것이라 생각해요
위 코멘트와 같은 맥락인데 static 변수들의 static 키워드를 제거하면 어떨까요?

testrace commented on src/main/java/nextstep/subway/entity/Sections.java:
```diff
@@ -45,6 +48,7 @@ public int getDistance() {
 
 	public void addSection(Section section) {
 		this.sections.add(section);
+		addPath(section.getDownStationId(), section.getUpStationId(), section.getDistance());
```
단순히 구간을 추가하는 경우 경로 조회를 하지 않을 수 있는데요
경로 조회와 상관 없이 매번 Path 객체에 역을 등록하기 보다 경로 조회시에 Path 객체를 생성해서 조회하면 어떨까요?

testrace reviewed (changes requested):
안녕하세요 휘연님 😃 
3단계 잘 구현하셨네요 👍 
몇 가지 코멘트 남겨두었는데 확인해 주시고 다시 리뷰 요청 부탁드려요

hhhhhwi committed:
refactor: 구간 추가와 상관없이 경로 조회시 PathFinder 객체를 생성하여 조회하도록 수정
- Path -> PathFinder로 클래스 명 변경
- 경로 조회 시 객체를 생성하므로 final, static 선언 제거
- Sections의 구간 추가, 제거 시 경로 수정하는 로직 삭제

hhhhhwi commented:
안녕하세요! 피드백 주신 대로 반영해보았습니다..!
제가 놓친 피드백이나 더 반영해야될 지점 있으면 코멘트 부탁드립니다 🔆 

감사합니다~

testrace commented on src/main/java/nextstep/subway/service/PathService.java:
```diff
@@ -13,17 +13,24 @@
 public class PathService {
 	private StationService stationService;
-	public PathService(StationService stationService) {
+	private SectionService sectionService;
```
final 을 추가하면 좋을 것 같아요
Service에서 Service를 의존하다보면 순환참조가 되는 경우가 발생할 수 있는데요
Repository를 참조하는 방향도 고려해 보시면 좋을 것 같아요

testrace commented on src/main/java/nextstep/subway/service/PathService.java:
```diff
@@ -13,17 +13,24 @@
 
 	public PathResponse getPath(Long source, Long target) {
-		List<Long> stations = Path.getPath(source, target);
+		PathFinder pathFinder = new PathFinder(sectionService.findAll());
+
+		return createPathResponse(pathFinder.getPath(source, target), (int) pathFinder.getDistance(source, target));
```
https://github.com/next-step/atdd-subway-path/pull/783#discussion_r1493883531
지난 코멘트에 대한 답변을 못드렸었네요 😢 

말씀해주신대로 서비스 단위 테스트에서 기능을 검증하면 도메인 단위 테스트와 크게 다를바가 없어서
서비스 단위 테스트의 필요성에 대해 고민하게 되는데요 (저도 그랬습니다 😃)
저는 서비스 단위 테스트에서는 mockito의 verify 메서드를 활용해 흐름(행위)를 검증하는 편입니다

testrace commented on src/test/java/nextstep/subway/unit/PathFinderTest.java:
```diff
@@ -0,0 +1,42 @@
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class PathFinderTest {
```
>인수테스트를 작성하고, 도메인을 설계 및 구현하는 방향인 Inside TDD 방식으로 사이클을 진행하다,
단위 테스트를 작성하며 api 구현을 진행하는 Outside TDD 방식으로 개발하려고 했으나..
해당 흐름으로는 제대로 진행을 못 했던 것 같습니다 ㅠ
>테스트를 먼저 작성을 하다보니, 구현을 억지로..? 맞춘다는 느낌도 없지 않았습니다.

양 방향 모두 경험을 해보는 것이 중요하다고 생각합니다 😃 
in-out 방식의 경우 도메인에 대한 지식이 갖춰지지 않은 상태라면 말씀하신대로 구현을 억지로 맞춘다는 느낌이 생기는 것 같아요
in-out 방식으로 구현하실 경우 도메인 설계를 먼저 하고 구현을 해보시면 좋을 것 같아요

testrace reviewed (approved):
안녕하세요 휘연님 😃 
피드백 깔끔하게 잘 반영해 주셨네요 👍 
이번 미션은 여기서 마무리하겠습니다 
몇 가지 코멘트 남겨두었는데 확인 부탁드려요
다음 미션도 화이팅입니다 🔥
